### PR TITLE
fix(compress): rank-gap delta guard (issue #45)

### LINKED ISSUE #45: /compress: rank-gap guard rejects genuine same-topic peers
## Problem

`/compress <topic>` Step 3.5 runs a vault search and applies two filters before offering to update an existing note:

```python
top["rank"] <= -5.0                                    # strong match
and abs(top["rank"]) > abs(results[1]["rank"]) * 1.5   # well ahead of #2
```

The 1.5× ratio guard fires **false-negative** when two or more indexed notes are all legitimately about the topic. Example from real usage:

- Query: `"PR 43 snapshot dev-test smoke results"`
- Top: `"PR #43 snapshot-summary dev-test — automated baseline (23/0/0)"` rank `-29.46`
- #2: `"Snapshot-summary integration — manual test plan (PR #43)"` rank `-24.71`
- `abs(-29.46) > abs(-24.71) × 1.5` → `29.46 > 37.06` → **False**
- Skill rejects the match and silently creates a duplicate

The guard punishes topics with accumulated notes (multi-phase PRs, iterative features) — exactly the topics that most benefit from consolidation.

## Location

`skills/compress/SKILL.md:88-92`

## Fix

Replace ratio guard with delta-score test:

```python
MIN_RANK_DELTA = 5.0  # tune against real rank distributions
rank_gap_ok = len(results) < 2 or (abs(top["rank"]) - abs(results[1]["rank"])) > MIN_RANK_DELTA
```

Delta-based guard scales naturally with rank magnitude — strong topics with several good matches still surface a clear leader without triggering spurious reject.

Tune `MIN_RANK_DELTA` by running `/compress <topic>` against the current vault for a handful of known-good and known-ambiguous queries; pick the value that splits them cleanly.

## Regression test

Manual: construct a fixture vault with two notes on the same topic (ranks ~-29 and ~-24), run the Step 3.5 query logic, assert `match: true` is returned.

## Source

Vault insight: `2026-04-18-compress-rank-gap-rejects-same-topic-peers-0cef.md`.

## Related

- Prior false-negative mode (different failure): `feedback_compress_search_false_negative.md` — off-topic notes outrank true matches on generic phrasing.

### PR DESCRIPTION
**Pull request by abhattacherjee**
## Summary

- Replaces the 1.5× ratio guard in `/compress` Step 3.5 with a delta-score guard (`|top.rank| - |#2.rank| > MIN_RANK_DELTA`), so genuine same-topic peer matches are no longer silently rejected as duplicates.
- Extracts the predicate into `hooks/compress_guard.py` so pytest and the tuning harness can import it directly.
- Ships a re-runnable tuning harness (`scripts/tune_compress_rank_gap.py`) + committed corpus (`scripts/compress_rank_gap_corpus.json`) so the threshold is auditable and re-tunable.

**Spec:** [github.com/abhattacherjee/spec-docs/.../2026-04-23-compress-rank-gap-delta-guard-design.md](https://github.com/abhattacherjee/spec-docs/blob/feature/issue-45-compress-rank-gap-peers/superpowers/specs/2026-04-23-compress-rank-gap-delta-guard-design.md)
**Plan:** [github.com/abhattacherjee/spec-docs/.../2026-04-23-compress-rank-gap-delta-guard-plan.md](https://github.com/abhattacherjee/spec-docs/blob/feature/issue-45-compress-rank-gap-peers/superpowers/plans/2026-04-23-compress-rank-gap-delta-guard-plan.md)

Closes #45.

## Empirical tuning

Ran the harness against the live vault. Found that vault growth since issue capture has compressed FTS5 rank deltas: the original issue's replay query (delta 4.75 at capture) now produces delta 0.78. The spec's stated feasible band (`[2.0, 4.5]`) no longer contains any threshold that fixes the bug against the live vault.

The spec's halt-and-revisit rule was triggered and followed. The tuning grid was extended downward and the spec's tuning protocol was updated (see `cbb7f2d` in the spec-docs repo). Shipped value: **`MIN_RANK_DELTA = 0.25`** (best feasible-band score 0.80).

### Tuning table

# Tuning sweep — 10 queries against /Users/abhishek/obsidian/claude-code-vault

## Per-query results

| id | query | top rank | #2 rank | delta | expected |
|---|---|---|---|---|---|
| replay-pr43-snapshot | `PR 43 snapshot dev-test smoke results` | -25.77 | -24.99 | 0.78 | True |
| synthetic-single-strong | `snapshot summary integration PR 43` | -15.48 | -14.99 | 0.49 | True |
| replay-compress-rank-gap-false-negative | `compress rank gap false negative same topic peers` | -44.68 | -15.67 | 29.01 | True |
| replay-hook-simulation-precompact | `hook simulation pattern precompact sessionend pipe JSON` | -40.64 | -11.19 | 29.44 | True |
| replay-recall-optimization-parallel-haiku | `recall optimization sprint parallel Haiku session performance` | -27.69 | -15.68 | 12.01 | True |
| replay-novel-unique-topic | `some novel unique topic xyz123` | -9.91 | -7.95 | 1.96 | False |
| synthetic-wide-gap | `SessionEnd hook atomic write temp rename vault note` | -22.26 | -20.17 | 2.09 | True |
| synthetic-narrow-gap-multi-match | `obsidian-brain security audit path containment hardening vulnerabilities` | -37.67 | -25.44 | 12.24 | True |
| synthetic-three-close-results | `vault search FTS5 sqlite index recall ranking score` | -33.34 | -31.08 | 2.26 | True |
| synthetic-off-topic-outrank | `fix the code check it` | -13.19 | -11.44 | 1.75 | False |

## Threshold sweep

| MIN_RANK_DELTA | TP | TN | FP | FN | score (TP+TN)/total | repro #45 |
|---|---|---|---|---|---|---|
| 0.25 | 8 | 0 | 2 | 0 | 0.80 | True |
| 0.50 | 7 | 0 | 2 | 1 | 0.70 | True |
| 0.75 | 7 | 0 | 2 | 1 | 0.70 | True |
| 1.00 | 6 | 0 | 2 | 2 | 0.60 | False |
| 1.50 | 6 | 0 | 2 | 2 | 0.60 | False |
| 2.00 | 6 | 2 | 0 | 2 | 0.80 | False |
| 3.00 | 4 | 2 | 0 | 4 | 0.60 | False |
| 4.00 | 4 | 2 | 0 | 4 | 0.60 | False |
| 5.00 | 4 | 2 | 0 | 4 | 0.60 | False |
| 6.00 | 4 | 2 | 0 | 4 | 0.60 | False |
| 7.00 | 4 | 2 | 0 | 4 | 0.60 | False |
| 8.00 | 4 | 2 | 0 | 4 | 0.60 | False |
| 10.00 | 4 | 2 | 0 | 4 | 0.60 | False |

## Selection guidance

- Hard constraint: MIN_RANK_DELTA must be < 4.75 so the issue #45
  repro case resolves True (see spec).
- Pick the feasible threshold with the best (TP+TN)/total score.
- Ties broken toward lower thresholds (more permissive = fewer
  silent duplicates; users can always decline the update prompt).

At the chosen threshold (0.25), 2 FPs remain (`replay-novel-unique-topic`, `synthetic-off-topic-outrank`). Both exemplify the off-topic-outrank failure mode explicitly listed as non-goal #2 in the spec — a user sees a wrong "update this note?" prompt and declines. Preferable to silent duplication (the issue #45 failure mode). A follow-up to address the outrank mode is already scoped in the spec's "Out of scope / follow-ups" section.

## Known pre-release gap

The installed plugin cache at `~/.claude/plugins/cache/.../obsidian-brain/2.4.1/hooks/` does not yet contain `hooks/compress_guard.py`. Until the next release ships:

- `/dev-test install` from this branch is the correct path for live smoke testing. The install script copies `hooks/*.py` so `compress_guard.py` lands in the cache.
- A live `/compress <topic>` invocation *without* `/dev-test install` will hit an `ImportError` which is now caught by a dedicated handler (see commit for this PR fixing the import-placement bug). The user-facing outcome is the same silent `{"match": false}` fallback, but stderr now shows an actionable message pointing at `/dev-test install`.

This matches spec Risk #3 (helper-import-from-SKILL.md risk class). Documented here for transparency.

## Test plan

- [x] `pytest tests/test_compress_rank_gap.py -v` → **11 passed** (boundary cases + repro lock + equal-ranks corner case + constant invariant)
- [x] `pytest tests/ -x` → **687 passed** (no regressions)
- [x] Direct Python smoke test with the issue #45 repro topic (`PR 43 snapshot dev-test smoke results`) returns `{"match": true, "path": .../2026-04-18-pr-43-snapshot-summary-dev-test-automated-baseli-e465.md, "rank": -25.64}` — the correct existing note
- [ ] Post-merge: one `/compress "PR 43 snapshot dev-test smoke results"` from a live session (after release) to confirm the update prompt surfaces

## Non-goals

- The absolute-strength threshold (`top.rank <= -5.0`) is unchanged.
- The off-topic-outranking false-negative (tracked as `feedback_compress_search_false_negative.md`) remains operator-side for now.
- No dev-test merge gate applies — this PR touches SKILL.md + a new library module, not hook lifecycle files. Memory rule `feedback_devtest_merge_gate` scopes that gate to hook-adjacent PRs.

### CODE REVIEW

File: hooks/compress_guard.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+Kept in its own module so the tuning harness (scripts/tune_compress_rank_gap.py)
+and pytest (tests/test_compress_rank_gap.py) can import it directly rather
+than shelling out through the SKILL.md embedded block.
+
+MIN_RANK_DELTA is empirically tuned against scripts/compress_rank_gap_corpus.json.
+Hard constraint: must be < 4.75 so the issue #45 repro case (top -29.46,
+runner-up -24.71) resolves to match: True. See the spec at
+docs/superpowers/specs/2026-04-23-compress-rank-gap-delta-guard-design.md
+for the full tuning protocol.
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/compress_guard.py]: ✅ Addressed in 3b9fa5e
**What was done:** Updated the docstring to make the external spec-docs repo explicit (`github.com/abhattacherjee/spec-docs` + the path within it), so readers don't expect a local `docs/` directory.

File: scripts/tune_compress_rank_gap.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        expected = entry["expected_match"]
+        if predicted and expected:
+            tp += 1
+        elif predicted and not expected:
+            fp += 1
+        elif not predicted and expected:
+            fn += 1
+        else:
+            tn += 1
+    return {"tp": tp, "tn": tn, "fp": fp, "fn": fn, "skipped": skipped}
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply scripts/tune_compress_rank_gap.py:64]: 💬 Kept the current `skipped` semantics as-is.

**Reasoning:** `skipped` is meant to distinguish "vault genuinely has no matches for this query" from "search could not be performed" (FTS5 error, locked DB, schema drift). Counting `not results` as TN/FN would conflate those two cases and mask infrastructure failures during re-tuning — a silent DB-locked failure would look like a batch of FNs and could push the threshold in the wrong direction.

The visibility half of your comment was already addressed in an earlier review (commit d249a59): the threshold-sweep table now includes a `skipped` column, and an all-empty diagnostic warns the user before the sweep if every query returned no results. So developers running the harness can see exactly how many entries were excluded and at which thresholds.

If a future corpus legitimately has many `expected_match: false` entries expected to return zero hits, we can revisit this and count those as TN. For now (`skipped = 0` across all rows in current runs), the conservative distinction is safer.

File: tests/test_compress_rank_gap.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
+"""Tests for hooks/compress_guard.py — /compress Step 3.5 rank-gap predicate.
+
+Covers boundary cases for `is_high_confidence_match()`, the pure predicate
+extracted from /compress SKILL.md Step 3.5. The helper is imported
+directly — no FTS5 fixture needed since the predicate takes a pre-sorted
+results list.
+"""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply tests/test_compress_rank_gap.py]: ✅ Addressed in 3b9fa5e
**What was done:** Removed the unused `import pytest`. Other test files in the repo have a mixed convention (some import but don't use it, some use pytest symbols); this file had no `pytest.*` calls so removing it is clean.

File: tests/test_compress_rank_gap.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+def test_empty_results_rejects():
+    assert is_high_confidence_match([]) is False
+
+
+def test_equal_ranks_rejects():
+    """Delta = 0 when top and runner-up have identical ranks.
+
+    FTS5 can produce identical rank values for near-identical notes. The
+    strict-> delta gate rejects (0 is never > any positive MIN_RANK_DELTA),
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply tests/test_compress_rank_gap.py]: ✅ Addressed in e4dac27
**What was done:** Rewrote the docstring fragment as `strict '>' delta gate` so it reads grammatically and doesn't look like an editing artifact.

File: skills/compress/SKILL.md [hunk 5]
```diff
@@ -84,17 +87,14 @@ try:
...
-        rank_gap_ok = len(results) < 2 or abs(top["rank"]) > abs(results[1]["rank"]) * 1.5
-        if top["rank"] <= -5.0 and rank_gap_ok:
-            print(json.dumps({"match": True, "path": top["path"], "title": top["title"], "date": top["date"], "tags": top["tags"], "rank": top["rank"]}))
-        else:
-            print(json.dumps({"match": False}))
+        print(json.dumps({"match": True, "path": top["path"], "title": top["title"], "date": top["date"], "tags": top["tags"], "rank": top["rank"]}))
     else:
         print(json.dumps({"match": False}))
+except ImportError as e:
+    print(f"Warning: plugin module missing ({e}) — run /dev-test install to apply the rank-gap fix", file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply skills/compress/SKILL.md]: ✅ Addressed in e4dac27
**What was done:** Dropped the "to apply the rank-gap fix" suffix; the handler now reads `plugin hooks are out of date or missing (...) — run /dev-test install`, which is accurate for `vault_index`, `obsidian_utils`, AND `compress_guard` ImportErrors.

File: tests/test_compress_rank_gap.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    results = [{"rank": -29.46}, {"rank": -24.71}]
+    assert is_high_confidence_match(results) is True
+
+
+def test_weak_top_rank_rejects():
+    # Top rank -3 fails the strength gate. The second element is also weak,
+    # so the list is properly sorted most-negative first. The strength gate
+    # fires before the delta gate is evaluated — no strong #2 is needed to
+    # prove the point.
+    results = [{"rank": -3.0}, {"rank": -4.0}]
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply tests/test_compress_rank_gap.py]: ✅ Addressed in 4409a5d
**What was done:** Swapped the test input from `[{-3.0}, {-4.0}]` to `[{-4.0}, {-3.0}]` so the list is properly sorted most-negative first (matching the predicate's documented contract). Updated the comment to explain that -4.0 is now the "top" at index 0 despite being weaker than -3.0 — this matches the FTS5 convention where more-negative = stronger match.

Good catch — this was the same sort-order violation class as the original `test_weak_top_rank_rejects_even_with_large_delta` bug; I swapped to the wrong direction in iteration 0's fix.
